singularity-eos/eos/eos_gruneisen.cpp

```diff
@@ -12,90 +12,233 @@
+PORTABLE_INLINE_FUNCTION Real find_min_bounded_val(const Real val1, const Real val2,
+                                                   const Real min, const Real max) {
+  // Try to find the minimum bounded value. If none exists, returns negative infinity
+  const Real minval = std::min(val1, val2);
+  const Real maxval = std::max(val1, val2);
+  if (minval > min && minval < max) {
+    // Minimum of two values is bounded so return it
+    return minval;
+  } else if (maxval > min && maxval < max) {
+    // Minimum isn't bounded but maximum is so return that
+    return maxval;
+  } else {
+    // Neither value is bounded so just return negative infinity to clearly show that the
+    // value is out of bounds
+    return -std::numeric_limits<Real>::infinity();
+  }
+}
+/*
+The Gruneisen EOS diverges at a specific compression. Ensure that the maximum density is
+below the smallest singularity in the reference pressure curve
+*/
+PORTABLE_FUNCTION Real Gruneisen::ComputeRhoMax(const Real s1, const Real s2, const Real s3,
+                                            const Real rho0) {
+  // Polynomial from the denominator of the reference pressure curve:
+  auto poly = [=](Real eta) { return 1 - s1 * eta - s2 * square(eta) - s3 * cube(eta); };
+
+  // First find the eta root. A negative root indicates that there is no maximum density.
+  Real root = 0; // Non-sensical root means a root hasn't been found yet
+  Real discriminant, root1, root2;
+  if (s2 == 0 && s3 == 0 && s1 > 0) {
+    // Linear Us-up analytic root
+    root = 1 / s1;
+  } else if (s3 == 0 && s2 != 0) {
+    // Quadratic Us-up
+    discriminant = square(s1) + 4 * s2;
+    if (discriminant < 0) {
+      root = -1; // Imaginary roots, so no limit
+    } else {
+      root1 = (s1 + std::sqrt(discriminant)) / (-2 * s2);
+      root2 = (s1 - std::sqrt(discriminant)) / (-2 * s2);
+      root = find_min_bounded_val(root1, root2, 0, 1);
+    }
+  } else if (s3 != 0) {
+    // Cubic Us-up: we'll use an iterative method to search for the minimum bounded root
+    // Note: when the discriminant of a cubic is less than zero, only one real root exists
+    //       so we don't need anything fancy to find the root.
+    discriminant = -18 * s3 * s2 * s1 - 4 * cube(-s2) + square(s2) * square(s1) -
+                   4 * s3 * cube(s1) - 27 * square(s3);
+    Real minbound = 0;
+    Real maxbound = 1;
+    if (discriminant >= 0) {
+      // Three real roots (roots may have multiplicity). We need to use the extrema to
+      // ensure we have a proper bracket in which to search for the root (if they exist).
+      // Note: If the discriminant is positive, then `square(s2) - 3 * s3 * s1` must
+      //       necessarily be positive (the reverse does not hold). Thus the extrema below
+      //       are not imaginary.
+      const Real extremum1 =
+          (2 * s2 + std::sqrt(square(-2 * s2) - 4 * 3 * s3 * s1)) / (-2 * 3 * s3);
```

Comment:
This is mostly for discussion, and am not really advocating for any change.

There could be some FLOPs saved here with a bit of simplification (may be wise to double-check me here)

```c++
ex1 = (s2 + std::sqrt(square(s2) - 3. * s3 * s1)/(-3. * s3);
```

and, if you wanna go REALLY crazy with it

```c++
auto threeS3 = 3. * s3;
auto I_threeS3 = 1. / threeS3; // division is expensive, so if there's reuse my instinct is to move to multipication
auto t1 = std::sqrt(square(s2) - threeS3 * s1);
Real ex1 = (s2 + t1) * I_threeS3;
Real ex2 = (s2 - t1) * I_threeS3;
```
This maybe a bit much, and likely as not most compilers are good enough in optimization here. There's also an argument to be made that in some contexts it's better to "load" FLOPS. On the other hand, keeping the FP arithmetic less "expressive" can help contain round-off and cancellation errors.

(side note: the quad formula is notorious for reaking havoc on precision, see Theorem 2 of https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html, https://www.johndcook.com/blog/2018/04/28/quadratic-formula/ for a nice summary).

Reply:
good catch

Reply:
> (side note: the quad formula is notorious for reaking havoc on precision, see Theorem 2 of https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html, https://www.johndcook.com/blog/2018/04/28/quadratic-formula/ for a nice summary).

I see @Yurlungur got the jump on me with this

Reply:
This code should be just executed at initialization so I'm not too concerned about performance here, BUT I think your first suggestion enhances readability since it's a pretty small simplification that doesn't make it any less obvious what is being calculated here.

I think the second suggestion makes the code slightly less readable though. The fact that you were able to instantly recognize the quadratic formula I think is a good thing here.

Reply:
FYI - I think I can get away with losing precision in the discriminants of the quadratic formulas here for two reasons:
1.  The value of the extrema calculations don't need to be really precise since they are used to bracket the real root more appropriately.
2.  When I'm directly calculating the roots for the quadratic case, I can reasonably rely on `s1` to be of order unity. That way although I may lose precision in the discriminant if it's near zero, I don't really care since I'll basically get two very close estimates of the root. If the lower estimate is too conservative or too affected by machine precision issues, then we can re-examine this issue later.